### PR TITLE
Add hostnameVerificationAlgorithm to reactive-datasource

### LINKED ISSUE #16484: reactive-pg-client sslMode=verify-full not working
## Describe the bug
When using reactive-pg-client it is not possible to use sslMode verify_full because it fails with:

```java
2021-04-13 17:23:24,498 ERROR [org.hib.rea.errors] (vert.x-eventloop-thread-3) could not execute query: java.util.concurrent.CompletionException: java.lang.IllegalArgumentException: Host verification algorithm must be specified under verify-full sslmode
	at java.base/java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:331)
	at java.base/java.util.concurrent.CompletableFuture.uniApplyNow(CompletableFuture.java:670)
	at java.base/java.util.concurrent.CompletableFuture.uniApplyStage(CompletableFuture.java:658)
	at java.base/java.util.concurrent.CompletableFuture.thenApply(CompletableFuture.java:2094)
	at java.base/java.util.concurrent.CompletableFuture.thenApply(CompletableFuture.java:143)
	at org.hibernate.reactive.pool.impl.ProxyConnection.withConnection(ProxyConnection.java:45)
	at org.hibernate.reactive.pool.impl.ProxyConnection.selectJdbc(ProxyConnection.java:109)
	at org.hibernate.reactive.loader.ReactiveLoader.executeReactiveQueryStatement(ReactiveLoader.java:129)
	at org.hibernate.reactive.loader.ReactiveLoader.doReactiveQueryAndInitializeNonLazyCollections(ReactiveLoader.java:69)
	at org.hibernate.reactive.loader.CachingReactiveLoader.doReactiveList(CachingReactiveLoader.java:62)
	at org.hibernate.reactive.loader.CachingReactiveLoader.reactiveListIgnoreQueryCache(CachingReactiveLoader.java:80)
	at org.hibernate.reactive.loader.custom.impl.ReactiveCustomLoader.reactiveList(ReactiveCustomLoader.java:94)
	at org.hibernate.reactive.session.impl.ReactiveSessionImpl.lambda$listReactiveCustomQuery$12(ReactiveSessionImpl.java:462)
	at java.base/java.util.concurrent.CompletableFuture.uniComposeStage(CompletableFuture.java:1106)
	at java.base/java.util.concurrent.CompletableFuture.thenCompose(CompletableFuture.java:2235)
	at java.base/java.util.concurrent.CompletableFuture.thenCompose(CompletableFuture.java:143)
	at org.hibernate.reactive.session.impl.ReactiveSessionImpl.listReactiveCustomQuery(ReactiveSessionImpl.java:462)
	at org.hibernate.reactive.session.impl.ReactiveSessionImpl.reactiveList(ReactiveSessionImpl.java:450)
	at org.hibernate.reactive.session.impl.ReactiveNativeQueryImpl.getReactiveResultList(ReactiveNativeQueryImpl.java:98)
	at org.hibernate.reactive.session.impl.ReactiveNativeQueryImpl.getReactiveSingleResult(ReactiveNativeQueryImpl.java:80)
	at io.smallrye.context.impl.wrappers.SlowContextualSupplier.get(SlowContextualSupplier.java:21)
	at io.smallrye.mutiny.operators.uni.builders.UniCreateFromCompletionStage.subscribe(UniCreateFromCompletionStage.java:24)
	at io.smallrye.mutiny.operators.AbstractUni.subscribe(AbstractUni.java:36)
	at io.smallrye.mutiny.operators.uni.UniRunSubscribeOn.lambda$subscribe$0(UniRunSubscribeOn.java:27)
	at org.hibernate.reactive.mutiny.impl.MutinySessionFactoryImpl.lambda$null$0(MutinySessionFactoryImpl.java:56)
	at io.vertx.core.impl.ContextImpl.executeTask(ContextImpl.java:366)
	at io.vertx.core.impl.EventLoopContext.lambda$executeAsync$0(EventLoopContext.java:38)
	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:164)
	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:472)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:500)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: java.lang.IllegalArgumentException: Host verification algorithm must be specified under verify-full sslmode
	at io.vertx.pgclient.impl.PgConnectionFactory.connect(PgConnectionFactory.java:141)
	at io.vertx.pgclient.impl.PgConnectionFactory.connectAndInit(PgConnectionFactory.java:105)
	at io.vertx.pgclient.impl.PgPoolImpl.connect(PgPoolImpl.java:47)
	at io.vertx.sqlclient.impl.ConnectionPool.check(ConnectionPool.java:207)
	at io.vertx.sqlclient.impl.ConnectionPool.acquire(ConnectionPool.java:76)
	at io.vertx.sqlclient.impl.PoolBase.getConnection(PoolBase.java:59)
	at io.quarkus.reactive.pg.client.runtime.ThreadLocalPgPool$PgPoolWrapper.getConnection(ThreadLocalPgPool.java:42)
	at io.quarkus.reactive.datasource.runtime.ThreadLocalPool.getConnection(ThreadLocalPool.java:104)
	at io.vertx.pgclient.PgPool_1d09fb569cbf9d284c8a4c30125d6bc55c456596_Synthetic_ClientProxy.getConnection(PgPool_1d09fb569cbf9d284c8a4c30125d6bc55c456596_Synthetic_ClientProxy.zig:116)
	at org.hibernate.reactive.pool.impl.SqlClientPool.lambda$getConnectionFromPool$1(SqlClientPool.java:76)
	at org.hibernate.reactive.pool.impl.Handlers.toCompletionStage(Handlers.java:24)
	at org.hibernate.reactive.pool.impl.SqlClientPool.getConnectionFromPool(SqlClientPool.java:75)
	at org.hibernate.reactive.pool.impl.SqlClientPool.getConnection(SqlClientPool.java:66)
	at org.hibernate.reactive.pool.impl.ProxyConnection.withConnection(ProxyConnection.java:44)
	... 28 more
```

But it is not possible to set host verification algorithm. We tried to set it as an uri parameter (hostnameVerificationAlgorithm) which is normally used by vertx but this does not work either.

### Expected behavior
You can configure hostnameVerificationAlgorithm the same way as you can configure sslMode.

### Actual behavior
You can only configure sslMode but not hostnameVerificationAlgorithm

## To Reproduce
No reproducer. It is sufficient to set sslMode without an existing db.

### Configuration
```properties
quarkus.datasource.db-kind=postgresql
quarkus.datasource.username=${POSTGRESQL_USER}
quarkus.datasource.password=${POSTGRESQL_PASSWORD}
quarkus.datasource.reactive.url=postgresql://${POSTGRESQL_HOST}:${POSTGRESQL_PORT}/${POSTGRESQL_DB}?hostnameVerificationAlgorithm=HTTPS
quarkus.datasource.reactive.trust-certificate-pem=true
quarkus.datasource.reactive.trust-certificate-pem.certs=${POSTGRESQL_SSL_ROOT_CERT}
quarkus.datasource.reactive.postgresql.ssl-mode=verify-full
```

### Output of `uname -a` or `ver`
Darwin LMUCM870527 19.6.0 Darwin Kernel Version 19.6.0: Tue Nov 10 00:10:30 PST 2020; root:xnu-6153.141.10~1/RELEASE_X86_64 x86_64

### Output of `java -version`
openjdk version "11.0.10" 2021-01-19 LTS
OpenJDK Runtime Environment Corretto-11.0.10.9.1 (build 11.0.10+9-LTS)
OpenJDK 64-Bit Server VM Corretto-11.0.10.9.1 (build 11.0.10+9-LTS, mixed mode)

### Quarkus version or git rev
1.13.1.Final

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)
3.6.3

### PR DESCRIPTION
**Pull request by markusdlugi**
Fixes #16484

### CODE REVIEW

**Comment by geoand** [comment]
CC @Sanne @gsmet

**Comment by Sanne** [comment]
awesome, thanks! 

Only one doubt: should we set a default hostname verification strategy when ssl-mode-full is set?

or maybe just check that one is set?

**Comment by markusdlugi** [comment]
There is already a check in the Vertx pgclient that a hostname verification algorithm needs to be set in case sslMode is verify-full. See #16484, which triggered this PR in the first place :)

About automatically setting a default algorithm in case we detect that the sslMode is verify-full: I was also wondering if that should be done, but decided against it because there are currently two different algorithms (HTTPS and LDAPS), so you would have to decide for one of those. And then you would also have a strange behavior for the default value - in case sslMode is verify-full, the default would be HTTPS (for example), but in case it's something else, the default would be "" (so no hostname verification). I thought that was rather confusing to have two different default values depending on the value of some other setting.

But if you think that should be done, I can also add such a behavior.

**Comment by Sanne** [comment]
> There is already a check in the Vertx pgclient that a hostname verification algorithm needs to be set in case sslMode is verify-full.

Yes that's true, but if I understood correctly that exception happens on the first attempt to run a query.

I think it would be best to fail to boot, and provide a clearer error immediately.

> About automatically setting a default algorithm in case we detect that the sslMode is verify-full: I was also wondering if that should be done, but decided against it because there are currently two different algorithms (HTTPS and LDAPS), so you would have to decide for one of those. And then you would also have a strange behavior for the default value - in case sslMode is verify-full, the default would be HTTPS (for example), but in case it's something else, the default would be "" (so no hostname verification). I thought that was rather confusing to have two different default values depending on the value of some other setting.
> But if you think that should be done, I can also add such a behavior.

Ok, no let's not switch the defaults automatically. I think you made a good point, and it would be probably confusing as well.

**Review by Sanne — CHANGES_REQUESTED** [review]
Looks good, but let's please add a check for the configuration to be valid at boot.

**Comment by markusdlugi** [comment]
The exception in #16484 actually happens at boot, because it fails to run the `SELECT 1` query to check the connection.
```
2021-04-15 11:43:39,550 ERROR [org.hib.rea.errors] (vert.x-eventloop-thread-3) failed to execute statement [SELECT 1]
```
But fair enough, it actually does not prevent the boot currently, Quarkus keeps running afterwards. I'll try to integrate such a check in the reactive DB clients.

**Comment by Sanne** [comment]
Thanks. Let's remember also that connections can be configured to be lazily initialized, so people wouldn't always have the error from `SELECT 1`.

File: extensions/reactive-datasource/runtime/src/main/java/io/quarkus/reactive/datasource/runtime/DataSourceReactiveRuntimeConfig.java [hunk 1]
```diff
@@ -109,6 +109,13 @@
     @ConfigItem(defaultValue = "PT1S")
     public Duration reconnectInterval = Duration.ofSeconds(1L);
 
+    /**
+     * The hostname verification algorithm to use in case the server's identity should be checked.
+     * Should be HTTPS, LDAPS or an empty string.
+     */
+    @ConfigItem
+    public Optional<String> hostnameVerificationAlgorithm = Optional.empty();
```
Comment (markusdlugi) [inline extensions/reactive-datasource/runtime/src/main/java/io/quarkus/reactive/datasource/runtime/DataSourceReactiveRuntimeConfig.java:117]: I was wondering if I should introduce an enum here? But all the other enums are typically already defined in Vertx. Here, the Vertx config is just a String, so we would need to define the enum ourselves in reactive-datasource.

Comment (tsegismont) [reply extensions/reactive-datasource/runtime/src/main/java/io/quarkus/reactive/datasource/runtime/DataSourceReactiveRuntimeConfig.java:117]: An enum wouldn't provide much, I think. The value depends on the type of client (e.g.g MySQL vs PostgreSQL).

**Review by tsegismont — APPROVED** [review]
LGTM, thank you @markusdlugi !

**Review by Sanne — APPROVED** [review]
Great! Thank you

**Pull request merged**